File: Disjoint Set/Java/ImplementationSize.java
// Time: O(4 alpha)
public class DisjointSetSize {
    int[] size;
    int[] ultimateParent;

    public DisjointSetSize(int num) {
        this.size = new int[num + 1];
        this.ultimateParent = new int[num + 1];

        Arrays.fill(size, 1);
        for(int index = 0; index <= num; index++) {
            this.ultimateParent[index] = index;
        }
    }

    private int findParent(int node) {
        if(this.ultimateParent[node] == node) {
            return node;
        }

        this.ultimateParent[node] = this.findParent(this.ultimateParent[node]);
        return this.ultimateParent[node];
    }

    private void unionBySize(int nodeOne, int nodeTwo) {
        int parentOne = this.findParent(nodeOne), parentTwo = this.findParent(nodeTwo);
        if(parentOne == parentTwo) {
            return;
        }
        int sizeOne = this.size[parentOne], sizeTwo = this.size[parentTwo];

        if(sizeOne < sizeTwo) {
            this.size[parentTwo] += sizeOne;
            this.ultimateParent[parentOne] = parentTwo;
        }
        else {
            this.size[parentOne] += sizeTwo;
            this.ultimateParent[parentTwo] = parentOne;
        }
    }

    public static void main(String[] args) {
        DisjointSetSize ds = new DisjointSetSize(7);
        int[][] edges = {{1, 2}, {2, 3}, {4, 5}, {6, 7}, {5, 6}, {3, 7}};

        for(int index = 0; index < edges.length - 1; index++) {
            ds.unionBySize(edges[index][0], edges[index][1]);
        }

        // Testing
        if(ds.findParent(3) == ds.findParent(7)) {
            System.out.println("Both are in same component");
        }
        else {
            System.out.println("Both are not in same component");
        }

        ds.unionBySize(3, 7);

        if(ds.findParent(3) == ds.findParent(7)) {
            System.out.println("Both are in same component");
        }
        else {
            System.out.println("Both are not in same component");
        }

    }
}

File: Disjoint Set/Python/ImplementationSize.py
# Time: O(4 alpha)
class disjointSetSize:
    
    def __init__(self, num):
        self.size = [1] * (num + 1)
        self.ultimateParent = [index for index in range(num + 1)]
    
    
    def findParent(self, node):
        if node == self.ultimateParent[node]:
            return node
        self.ultimateParent[node] = self.findParent(self.ultimateParent[node])
        return self.ultimateParent[node]
    
    def unionBySize(self, nodeOne, nodeTwo):
        parentOne, parentTwo = self.findParent(nodeOne), self.findParent(nodeTwo)
        
        if parentOne == parentTwo:
            return
    
        sizeOne, sizeTwo = self.size[parentOne], self.size[parentTwo]
        
        if sizeOne < sizeTwo:
            self.ultimateParent[parentOne] = parentTwo
            self.size[parentTwo] += sizeOne
        else:
            self.ultimateParent[parentTwo] = parentOne
            self.size[parentOne] += sizeTwo

ds = disjointSetSize(7)
edges = [[1, 2], [2, 3], [4, 5], [6, 7], [5, 6]]

for one, two in edges:
    ds.unionBySize(one, two)
    
if ds.findParent(3) == ds.findParent(7):
    print("Both are in same component")
else:
    print("Both are not in same component")

ds.unionBySize(3, 7)

if ds.findParent(3) == ds.findParent(7):
    print("Both are in same component")
else:
    print("Both are not in same component")
        
